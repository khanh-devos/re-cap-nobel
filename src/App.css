@import url("https://fonts.googleapis.com/css?family=Lato");

:root {
  --color1: #ec4c8b;
  --color2: #fe5395;
  --color3: #cf4278;
  --stdPadd: 0.2rem;
  --imgHeight: 5rem;
  --padMobile: 15rem;
  @media screen and (max-width: 768px) {
    --padMobile: 0.2rem;
  }

}

body {
  font-size: 14px;
  font-family: Lato, fantasy;
  margin: auto;
  padding: 0.2rem var(--padMobile);
  background: var(--color1);
}

* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
  font-family: Lato, fantasy;
  color: #fff;
}

.nation {
  min-height: 100%;
  margin: auto;
}

.header-top-line {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: var(--stdPadd);
  padding-left: 0;
  background: var(--color1);
}

.header-back-btn {
  width: 1.3rem;
  border: none;
  font-size: 1.2rem;
  background: var(--color1);
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
}

.header-back-btn:active {
  background: red;
}

.header-icon-container {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  width: 3rem;
}

.header-grid {
  display: grid;
  grid-template-columns: 1fr 1fr;
  background: var(--color2);
  height: 7rem;
}

.header-image {
  height: var(--imgHeight);
  max-width: 100%;
  padding: var(--stdPadd);
  margin: auto;
  filter: opacity(0.3);
}

.header-location {
  background: var(--color2);
  padding: 0 10%;
  display: flex;
  flex-direction: column;
  justify-content: center;
}

.header-stats {
  background: var(--color1);
  padding: var(--stdPadd);
}

.nation-grid-container {
  display: grid;
  grid-template-columns: 1fr 1fr;
}

.nation-item {
  position: relative;
  height: 9.5rem;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  align-items: flex-end;
  padding: 0.5rem;
  padding-bottom: 1rem;
  border: none;
  background: var(--color3);
}

.nation-item:active {
  transform: scale(1.1);
  z-index: 2;
}

.nation-item-img {
  max-width: 100%;
  height: var(--imgHeight);
  margin: auto;
  filter: opacity(0.2);
}

.nation-item-circle-right-arrow-icon {
  position: absolute;
  top: var(--stdPadd);
  right: var(--stdPadd);
  opacity: 0.8;
}

.nation-item-text {
  text-align: right;
}

.header-bolden {
  font-weight: bold;
}

/* NATION ROUTES */
.nation-route-flex {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
}

.nation-route-item {
  height: 5rem;
  padding: var(--stdPadd);
  border: none;
  background: var(--color2);
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
}

.nation-route-item:nth-child(even) {
  background: var(--color3);
}

.nation-route-item-part2 {
  display: flex;
  flex-direction: row;
  gap: 1rem;
}

.winners {
  padding: var(--stdPadd);
}
