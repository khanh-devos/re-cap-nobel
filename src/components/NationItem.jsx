import { PropTypes } from 'prop-types';
import { useNavigate } from 'react-router-dom';

export default function NationItem({ country, count }) {
  const navigate = useNavigate();
  const handleClick = (event) => {
    event.preventDefault();
    navigate(`${event.target.id}`);
  };

  return (
    <button type="button" id={`${country}`} onClick={handleClick}>
      {country}
      {' '}
      :
      {' '}
      {count}
    </button>
  );
}

NationItem.propTypes = {
  country: PropTypes.string.isRequired,
  count: PropTypes.number.isRequired,
};
