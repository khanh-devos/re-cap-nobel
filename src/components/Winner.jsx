import { PropTypes } from 'prop-types';
import { useSelector } from 'react-redux';

export default function Winner({ id }) {
  const { nobels } = useSelector((store) => store.nobel);
  const {
    fullname, city, date, awardYear,
  } = nobels.find((item) => item.id === id);

  return (
    <div>

      <h3>
        Name :
        {fullname}
      </h3>
      <h3>
        City :
        {city}
      </h3>
      <h3>
        BirthDate :
        {date}
      </h3>
      <h3>
        Awarded Year :
        {awardYear}
      </h3>
      <p>--------------------</p>
    </div>
  );
}

Winner.propTypes = {
  id: PropTypes.string.isRequired,
};
