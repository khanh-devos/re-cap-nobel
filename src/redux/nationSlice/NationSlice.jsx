import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';
import axios from 'axios';

export const API_NOBEL = 'https://api.nobelprize.org/2.1/laureates';

const initialState = {
  nobels: [],
  isLoading: true,
  amount: 0,
  uniqueCountries: [],
};

export const fetchNobel = createAsyncThunk(
  'nobelSlice/fetchNobel',
  async (name, thunkAPI) => {
    try {
      const res = await axios.get(API_NOBEL);
      return res.data;
    } catch (err) {
      return thunkAPI.rejectWithValue('st went wrong');
    }
  },
);

const takeUniqueCountries = (arr) => {
  const uniqueCountries = [];
  const compare = [];

  arr.forEach(({ country, id, city }) => {
    if (!compare.includes(country)) {
      compare.push(country);
      uniqueCountries.push({ id, country, cities: [city] });
    } else {
      const check = uniqueCountries.find((item) => item.country === country);
      check.cities.push(city);
    }
  });
  return uniqueCountries;
};

const nobelSlice = createSlice({
  name: 'nobelSlice',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchNobel.pending, (state) => ({
        ...state,
        isLoading: false,
      }))
      .addCase(fetchNobel.fulfilled, (state, { payload }) => {
        const arr = payload.laureates.map((item) => {
          const { id, fullName: { en: fullname }, gender } = item;
          const { date } = item.birth;
          const { en: city } = item.birth.place?.cityNow || { en: 'Unknown_City' };
          const { en: country } = item.birth.place?.countryNow || { en: 'Unknown_Country' };

          const {
            awardYear,
            categoryFullName: { en: category },
            prizeAmount: prize,
          } = item.nobelPrizes[0];

          // const city = city123.replace(/[^a-z0-9]/gi, '_').toLowerCase();
          // const country = country123.replace(/[^a-z0-9]/gi, '_').toLowerCase();

          return {
            id, fullname, gender, date, city, country, awardYear, category, prize,
          };
        });

        return {
          ...state,
          isLoading: false,
          nobels: arr,
          amount: arr.length,
          uniqueCountries: takeUniqueCountries(arr),
        };
      })
      .addCase(fetchNobel.rejected, (state) => ({
        ...state,
        isLoading: false,
      }));
  },
});

export default nobelSlice.reducer;
