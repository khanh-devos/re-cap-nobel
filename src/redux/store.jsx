import { configureStore } from '@reduxjs/toolkit';
import NobelReducer from './nationSlice/NationSlice';

const store = configureStore({
  reducer: {
    nobel: NobelReducer,
  },
});

export default store;
