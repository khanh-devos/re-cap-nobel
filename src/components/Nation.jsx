import { useSelector } from 'react-redux';

import NationItem from './NationItem';

export default function Nation() {
  const { uniqueCountries } = useSelector(((store) => store.nobel));

  return (
    <div>
      {
      uniqueCountries.map(({ id, country, cities }) => (
        <NationItem
          key={id}
          country={country}
          count={cities.length}
        />
      ))
      }
    </div>
  );
}
