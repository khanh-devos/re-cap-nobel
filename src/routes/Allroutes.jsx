import { useSelector } from 'react-redux';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { v4 as uuidv4 } from 'uuid';

import Nation from '../components/Nation';
import CityRoute from './CityRoute';
import NationRoute from './NationRoute';

export default function Allroutes() {
  const { uniqueCountries } = useSelector((store) => store.nobel);

  return (
    <BrowserRouter basename="re-cap-nobel">
      <Routes>
        <Route path="/" element={<Nation />} />
        {
            uniqueCountries.map(({ id, country, cities }) => {
              const r1 = [
                <Route
                  key={uuidv4()}
                  path={`/${country}`}
                  element={<NationRoute country={country} />}
                />,
              ];

              cities.forEach((city) => (
                r1.push(<Route
                  key={uuidv4()}
                  path={`/${country}/${city}`}
                  element={<CityRoute city={city} country={country} />}
                />)));

              return r1;
            })
        }
      </Routes>
    </BrowserRouter>
  );
}
