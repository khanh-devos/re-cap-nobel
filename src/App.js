import { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate } from 'react-router-dom';

import './App.css';
import { fetchNobel } from './redux/nationSlice/NationSlice';
import Allroutes from './routes/Allroutes';

function App() {
  const dispatch = useDispatch();

  useEffect(() => {
    dispatch(fetchNobel());
  }, [dispatch]);

  return (
    <div className="App">
      <Allroutes />
    </div>
  );
}

export default App;
