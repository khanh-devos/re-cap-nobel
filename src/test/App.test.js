import { render, screen, act } from '@testing-library/react';
import axios from 'axios';
import { Provider } from 'react-redux';
import { API_NOBEL } from '../redux/nationSlice/NationSlice';
import store from '../redux/store';
import App from '../App';
import Allroutes from '../routes/Allroutes';
import CityRoute from '../routes/CityRoute';
import NationRoute from '../routes/NationRoute';
import Winner from '../components/Winner';
import { useAppNavigate, useAppSelector } from './mock';
import Nation from '../components/Nation';
import NationItem from '../components/NationItem';
import { MemoryRouter, Route, Router, Routes } from 'react-router-dom';

jest.mock('axios');

describe('Test all the App:', () => {
  const data = {
  laureates: [
    {
      id: '1',
      fullName: {
        en: 'A. Michael Spence',
        se: 'A. Michael Spence',
      },
      gender: 'male',
      birth: {
        date: '1943-00-00',
        place: {
          cityNow: {
            en: 'city1',
          },
          countryNow: {
            en: 'c1',
          },
        },
      },
      nobelPrizes: [
        {
          awardYear: '2001',
          categoryFullName: {
            en: 'The Sveriges Riksbank Prize in Economic Sciences in Memory of Alfred Nobel',
          },
          prizeAmount: 10000000,
        },
      ],
    },
    {
      id: '2',
      fullName: {
        en: 'Khanh Dom',
      },
      gender: 'male',
      birth: {
        date: '1980-00-00',
        place: {
          cityNow: {
            en: 'city2',
          },
          countryNow: {
            en: 'c2',
          },
        },
      },
      nobelPrizes: [
        {
          awardYear: '2001',
          categoryFullName: {
            en: 'The Sveriges Riksbank Prize in Economic Sciences in Memory of Alfred Nobel',
          },
          prizeAmount: 10000000,
        },
      ],
    },

    ]
  }

  beforeEach(() => {
    axios.get.mockImplementation((url) => {
      switch (url) {
        case API_NOBEL:
          return Promise.resolve({ data });
        default:
          return Promise.reject(new Error('not found'));
      }
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match the snapshot APP', async () => {
    await act(async () => {
      const { container } = render(
        <Provider store={store}>
          <App />
        </Provider>,
      );
      expect(container).toMatchSnapshot();
    });
  });

  it('should match the snapshot Allroutes', async () => {
    await act(async () => {
      const { container } = render(
        <Provider store={store}>
          <Allroutes />
        </Provider>,
      );

      expect(container).toMatchSnapshot();
    });
  });

  it('should match the snapshot CityRoute', async () => {
    await act(async () => {
      const { container } = render(
      <Provider store={store}>
        <MemoryRouter initialEntries={["/c2"]}>
          <Routes>
            <Route path="c2" element={<CityRoute country={'c2'} city={'city2'}/>} />
          </Routes>
        </MemoryRouter>
      </Provider>,
      );
      expect(container).toMatchSnapshot();
    });
    
  });

  it('should match the snapshot NationRoute', async () => {
    await act(async () => {
      const { container } = render(
        <Provider store={store}>
          <NationRoute country={'c2'}/>
        </Provider>,
      );
      expect(container).toMatchSnapshot();
    });
  });

  it('should match the snapshot Nation', async () => {
    await act(async () => {
      const { container } = render(
        <Provider store={store}>
          <Nation />
        </Provider>,
      );
      expect(container).toMatchSnapshot();
    });
  });

  it('should match the snapshot NationItem', async () => {
    await act(async () => {
      const { container } = render(
        <Provider store={store}>
          <NationItem country={'c1'} count={2} />
        </Provider>,
      );
      expect(container).toMatchSnapshot();
    });
  });

  it('should match the snapshot Winner', async () => {
    await act(async () => {
      const { container } = render(
        <Provider store={store}>
          <Winner id={'1'} />
        </Provider>,
      );
      expect(container).toMatchSnapshot();
    });
  });

});
