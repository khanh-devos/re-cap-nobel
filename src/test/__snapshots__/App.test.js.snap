// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test all the App: should match the snapshot APP 1`] = `<div />`;

exports[`Test all the App: should match the snapshot APP 2`] = `<div />`;

exports[`Test all the App: should match the snapshot Allroutes 1`] = `<div />`;

exports[`Test all the App: should match the snapshot CityRoute 1`] = `<div />`;

exports[`Test all the App: should match the snapshot Nation 1`] = `<div />`;

exports[`Test all the App: should match the snapshot NationItem 1`] = `<div />`;

exports[`Test all the App: should match the snapshot NationRoute 1`] = `<div />`;

exports[`Test all the App: should match the snapshot Winner 1`] = `<div />`;

exports[`Test rejection case should match the snapshot APP 1`] = `<div />`;
